flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -728,6 +766,17 @@ private void persistAndRunJob(ExecutionPlan executionPlan) throws Exception {
 
     private JobManagerRunner createJobMasterRunner(ExecutionPlan executionPlan) throws Exception {
         checkState(!jobManagerRunnerRegistry.isRegistered(executionPlan.getJobID()));
+
+        JobStatusListener singleJobApplication = null;
+        Optional<AbstractApplication> optionalApplication =
+                executionPlan.getApplicationId().map(applications::get);
+        if (optionalApplication.isPresent()) {
+            AbstractApplication application = optionalApplication.get();
+            if (application instanceof JobStatusListener) {
```

Comment:
I think the clearer logic should be: if an application is a `SingleJobApplication`, it needs be passed to the jobManager as a JobStatusListener.